### ISSUE TITLE
[Feature]: desketchify the out.js

### ISSUE THREAD

**Issue by mvsite** [movie-web/movie-web#264]
### What feature do you want to add?

It's be useful to have a de-sketchified out.js.  The first thing I did with this software is remove dependencies on remote content so it was completely self-hosted, and right away noticed the out.js was encoded and obfuscated, which I was able to deobfuscate considerably but there is still a lot of non-obvious functionality including what appears to be reading of keystrokes, debugger, etc and maybe it's all harmless but the obfuscation begs the question: why try to hide the code?

### Why do you want to have this feature?

Well, it would be easier to trust the code if it wasn't obfuscated.

**Comment by zisra** [comment]
To hide the cors proxy URL which the developers have to pay for

**Comment by mvsite** [comment]
This is bogus.  Here's part of why:

1. the cors url of theirs is trivially discoverable by reverse engineering the javascript, or by just sniffing the traffic while using their app:
proxy-02.no-cors-proxy.workers.dev
cors.movie-web.app

2. both of those can be publicly written because they are useless to anyone but legitimate users - they have an ACL that prevents other sites accessing them.  I tested to make sure, before posting them here to make the point, and they refuse any proxy behavior.

3. probably most importantly, IF you really thought this "protection" was needed then wtf would you design the app like this so every person who runs it has THEIR cors readily exposed in /config.js open to the world?  The live https://movie-web.app/ website doesn't have any cors in it's /config.js, yet your instructions (and in fact the app's requirement) is to have our cors in that file, easily readable by anyone.  I had a test server up, which you can see because you snuck in google analytics snitch telemetry and I didn't think to block that the first time, and have since deleted it all and am just making my own instead.

4. there are free cors services such as https://corsproxy.io/ that are wide open and usable, so there is no need for all this.

Bottom line is the enciphered script is doing sketchy behavior with keystroke logging and anti-debugger (for those who only know basic built-in f12 technique) and is unnecessary and should be replaced.  If your only excuse is "cors" then that's blown up and wrong for all the reasons I said above.

**Comment by mrjvs** [comment]
Hey @mvsite, thanks for your interest in movie-web.
I understand your concern with the obfuscated code, it does seem very sketchy in an open source project.

What zisra said is partially correct, the script (which we call anti-tamper internally) does a best-effort attempt at preventing misuse of our proxy.

To counter the argument that we can just use free services instead, we handle a million page views a month. That means we get a lot of proxy requests, about 200k a day. we don't want to overload any public cors proxies. Not to mention that because it circumvents SSL, it's generally unsafe to use public cors proxies. The public proxies could be collecting all kinds of data, maybe even session cookies if we ever have something like it.

#### So why do we have the script?

We've had problems where people would take our cors proxy and spam it, causing the entire app to not function for months. This is the entire reason script exists, we wouldn't have done it if people didnt do malicious things with it.

They didn't just selfhost, they sent requests for their own app and maxed out the usage.

One solution was to obfuscate everything and not be open source, however we really like open source software and want to stay open source.

The other solution was the anti-tamper script, however we realized that would make it pretty much impossible to selfhost.
We choose to make the anti-tamper script run only with a hidden config value, that means the anti-tamper script simply doesn't run on selfhosted instances and on local development servers. We also made efforts to make selfhosting easier: we made a selfhosting guide, we made a runtime config (config.js, compile-time is possible with .env files) and made pre-compiled releases in github for everyone to download.

Also keep in mind that this is a client only application, it does not have a backend. That means that **anything** stored in any configuration or constant - even without using `config.js` - cannot be hidden by definition.

#### What does the script do?

The anti-tamper script does two things:
 1. Make it more difficult to use dev tools in production. This is to prevent sniffing the proxy url in the network tab.
 2. Modify the global fetch object, it does some magic with the proxy requests, if that magic isnt done then the proxy refuses the request.

Yes, the proxy can only be used if you have the `Host` header set correctly. But if that was the only protection, then people could still use the proxy on backend services. So we added the request magic to basically make the entire thing useless for anyone except the original hoster.

Truth be told, if someone really wishes to take us down, they could just spam the requests once they bypass the anti-devtools behaviour to find the urls (which is really easy if you know how, as you said). However our anti-tamper scripts prevents them from using it for anthing.

---

A few final notes:
 - The script isn't malicious nor is it hard to reverse engineer. In fact somebody in the discord server already deobfuscated the entire code manually in less than an hour, and thought nothing of the contents.
 - If you don't like google analytics in your instance, change the analytics ID in `src/setup/constants.ts`
 - If we were to unobfusticate the anti-tamper script, all the reasons why we have it in the first place would be voided. Making it as good as having no protection at all.
 - We know it isn't going to stop everyone, but as said before, this is a best-effort attempt to stop malicious misuse. Which it has stopped successfully thus far.
 
We don't like having this either, but we saw this as the best way to stop people from misuing our proxy. It's been working wonders so far, we haven't had anyone maliciously using our proxy after these changes (hopefully didnt jinx it).
 
If you don't want the anti-tamper script and the google analytics. You are free to make fork.
You can remove google analytics code in `/src/setup/ga.ts` and the out.js script tag can be removed in `/index.html`.

I'm gonna close the issue, however if you have any remaining questions, feel free to ask.

**Comment by mvsite** [comment]
Excellent answer, thanks for clearing that up.   (btw I was going to lazily use vidsrc.me for every video but you guys put in some work and maybe I should just use your solution.  thanks again).
